pytorch/imagerecognition/cifar10-resnet20-distributed-data-parallel/main.py

```diff
@@ -0,0 +1,289 @@
+
+        tracker = Tracker(metrics, run_id, rank, goal=goal)
+
+        dist.barrier()
+
+        tracker.start()
+
+        for epoch in range(0, train_epochs):
+            train_round(
+                train_loader,
+                model,
+                optimizer,
+                loss_function,
+                metrics,
+                scheduler,
+                "fp32",
+                schedule_per="epoch",
+                transform_target_type=None,
+                use_cuda=use_cuda,
+                max_batch_per_epoch=max_batch_per_epoch,
+                tracker=tracker,
+            )
+
+            is_best = validation_round(
+                val_loader,
+                model,
+                loss_function,
+                metrics,
+                run_id,
+                rank,
+                "fp32",
+                transform_target_type=None,
+                use_cuda=use_cuda,
+                max_batch_per_epoch=max_batch_per_epoch,
+                tracker=tracker,
+            )
+
+            checkpointer.save(
+                tracker, model, optimizer, scheduler, tracker.current_epoch, is_best
+            )
+
+            tracker.epoch_end()
+
+            if tracker.goal_reached:
+                print("Goal Reached!")
+                time.sleep(10)
+                return
+
+    else:
+        cecf = CheckpointsEvaluationControlFlow(
+            ckpt_dir=ckpt_run_dir,
+            rank=rank,
+            world_size=world_size,
+            checkpointer=checkpointer,
+            model=model,
+            epochs=train_epochs,
+            loss_function=loss_function,
+            metrics=metrics,
+            use_cuda=use_cuda,
+            dtype="fp32",
+            max_batch_per_epoch=None,
+        )
+
+        train_stats = cecf.evaluate_by_epochs(train_loader)
+        with open(os.path.join(output_dir, "train_stats.json"), "w") as f:
+            json.dump(train_stats, f)
+
+        val_stats = cecf.evaluate_by_epochs(val_loader)
+        with open(os.path.join(output_dir, "val_stats.json"), "w") as f:
+            json.dump(val_stats, f)
+
+
+def main(
+    run_id,
+    dataset_dir,
+    ckpt_run_dir,
+    output_dir,
+    rank,
+    hosts,
+    backend,
+    validation_only=False,
+    gpu=False,
+    light_target=False,
+):
+    r"""Main logic."""
+
+    with initialize_backends(
+        comm_backend=backend,
+        hosts=hosts,
+        rank=rank,
+        logging_level="INFO",
+        logging_file=os.path.join(output_dir, "mlbench.log"),
+        use_cuda=gpu,
+        seed=42,
+        cudnn_deterministic=False,
+        ckpt_run_dir=ckpt_run_dir,
+        delete_existing_ckpts=not validation_only,
+    ):
+        train_loop(
+            run_id,
+            dataset_dir,
+            ckpt_run_dir,
+            output_dir,
+            validation_only,
+            use_cuda=gpu,
+            light_target=light_target,
+        )
+
+
+if __name__ == "__main__":
```

Comment:
is this needed in each main.py for each implementation or should this be provided somewhere by mlbench core?

Reply:
I am not sure what you mean. We need to parse the arguments inside the benchmark.

Reply:
Could be looked into in https://github.com/mlbench/mlbench-core/issues/85